### PR TITLE
Use GEOS >= 3.9.0 fixed precision

### PR DESCRIPTION
**Pull request by lbartoletti**
## Description

Since GEOS 3.9.0 it's possible to use a fixed precision to perform some operations using the OverlayNG engine.

### API changes
I added a parameter gridSize with a default value at -1. Since GEOS requires a value > 0, it falls back to the normal function if the condition is not satisfied, and so the actual behavior.

### Processing
In this PR, only the Processing will use this new parameter. I exposed a new optional "precision" field in:
- Overlaps, Intersection, Difference, SymDifference, Union

image

Sponsored by: Orange Telecom FR

### CODE REVIEW

**Comment by andreasneumann** [comment]
Nice!

How is this exposed to the user? An additional grid size parameter?

Can we derive the gridSize value from the "Geometry precision" in the digitzing properties of a layer?

image

We'd have to define some logic:

if multiple layers are involved, the precision value would be taken from the "coarsest" value by default, if only one of the layers has it defined, it would rule?

**Comment by lbartoletti** [comment]
@andreasneumann thanks,

> How is this exposed to the user? An additional grid size parameter?

At the moment (and for this PR), it's only an API change. A new parameter is exposed.

I also updated some processing algorithm, see my edited description with a screenshot.
It's not intended to be used in the digitzing panel (maybe later ?).

**Comment by DelazJ** [comment]
@lbartoletti Can you add a help to the parameter, please? I think we should consider adding a help in the tooltip for any new parameter in Processing from now.
And also wondering whether this parameter belongs to the advanced group or no? But it's unclear to me what it actually does. For the moment!

File: src/core/geometry/qgsgeometryengine.h
Comment (nyalldawson) [inline src/core/geometry/qgsgeometryengine.h:113]: The QgsGeos constructor already has a "precision" argument -- there's no need for an additional argument for each of these methods

File: src/core/geometry/qgsgeometry.h
Comment (nyalldawson) [inline src/core/geometry/qgsgeometry.h:1703]: The gridSize arguments need documentation, including a "since QGIS 3.24" and a description of what they mean

**Comment by nyalldawson** [comment]
Personally I'm -1 to exposing precision in the processing algorithms. We used to have a similar option in QGIS 2.x, and it was a cause of much user confusion. Users expected it to mean "within this distance tolerance", and not "results will be completely unpredictable/random at distances less than this tolerance". I honestly doubt that there's a real need for an end user to ever need to set this.

If there IS a need to expose this in an algorithm, can I suggest it's done instead in a plugin which exposes "relations with tolerance" algorithms instead?

**Comment by andreasneumann** [comment]
> Personally I'm -1 to exposing precision in the processing algorithms. We used to have a similar option in QGIS 2.x, and it was a cause of much user confusion. Users expected it to mean "within this distance tolerance", and not "results will be completely unpredictable/random at distances less than this tolerance". I honestly doubt that there's a real need for an end user to ever need to set this.
> 
> If there IS a need to expose this in an algorithm, can I suggest it's done instead in a plugin which exposes "relations with tolerance" algorithms instead?

Here a have to disagree. On the contrary. There is a lot of need for these parameters.

Since the gridSize parameter was introduced in Postgis/GEOS 3.9, our SQL analysis code runs more robust and more predictable. We need this parameter in QGIS as well!

See http://blog.cleverelephant.ca/2020/12/waiting-postgis-31-3.html (fixed precision overlay).

With the fixed precision overlay, things run much better for us!

**Comment by lbartoletti** [comment]
> Here I have to disagree. On the contrary. There is a lot of need for these parameters.
> 
> Since the gridSize parameter was introduced in Postgis/GEOS 3.9, our SQL analysis code runs more robust and more predictable. We need this parameter in QGIS as well!
> 
> 
> With the fixed precision overlay, things run much better for us!

I can't say it better, that's exactly why we want to offer this feature in QGIS.

**Comment by lbartoletti** [comment]
@DelazJ 
> Can you add a help to the parameter, please?

Yes, it's still a draft. I wanted to check if the CI was ok (difference between the CI and my workstation) to continue the dev.

> I think we should consider adding a help in the tooltip for any new parameter in Processing from now.

OK, I'll do it too.

> And also wondering whether this parameter belongs to the advanced group or no? 

Yes, I also think it should go into the advanced group.

**Comment by nyalldawson** [comment]
> With the fixed precision overlay, things run much better for us!

Hmm, in that case there's a few things I'd ask/propose:

- Can we word the parameters in such a way that a user will NOT mistake them for a "tolerance" value? Maybe we could make the parameter name more technical sounding in order to detract from beginner users from messing with it. 
- It definitely needs to be marked advanced, and tooltips/documentation added for all the algorithms which expose it explaining it's use
- Question: would it be better to expose this through the processing context instead of via individual algorithms? If you're running a model, I can imagine that many steps in the model will need to have the size set and will likely all need to be the same size. Instead of adding it as a parameter to multiple algorithms, we could instead make it a general property in QgsProcessingContext and allow users to set this value via a settings panel in the algorithm dialog (I've wanted to add something like this for a while to allow users to control other context properties, like the area/distance units and ellipsoid settings). We could potentially even do both approaches, and make sure the size parameter in the algorithms is optional and defaults to not set, which means "use the default size" (aka take the size from the context).  Then we'd have a way for a model to have a specific size for just one step if needed, while still allowing users to modify the size for the model in general in one place (so that it can be set to appropriate values for the actual inputs being used at run time).

**Comment by andreasneumann** [comment]
> * Question: would it be better to expose this through the processing context instead of via individual algorithms? If you're running a model, I can imagine that many steps in the model will need to have the size set and will likely all need to be the same size.

Agreed.  A more "central" way to define this parameter would be useful.

In our use cases it its usually a millimeter precision that we want to target globally. Even across all projects.

A global and/or per project setting would be useful.

We could also use a variable for that purpose, that already allows to have different scopes (global, project, layer)

**Comment by nyalldawson** [comment]
@andreasneumann 

> A global and/or per project setting would be useful.

We could have a "global" default setting in the Options -> Processing tab, and then allow users to modify this on a per-run basis through a "settings" pane in the algorithm dialog.

**Comment by strk** [comment]
Could a PR be marked in a way to avoid stale bot from closing it ?
There's some work in here, and some discussion, the recent comments are all a noisy war between the stale bot and humans...

**Comment by strk** [comment]
I think closing work which is unfinished just moves out of the radar the presence of possible partners for reaching a goal. Someone asked for this kind of functionality on IRC today, and I did remember this work, went looking and did not find it easily, just because it was closed. But "closed" not because merged, just because a bot decided ideas have to vanish unless implemented. I find this a bit violent and discouraging for contributors.

I see in history that @lbartoletti removed the "stale" label on Jan 6, Feb 20, March 7, March 29. That's 4 times. Stale bot just had more power into doing its thing (interrupting this effort) and @lbartoletti understandably gave up fighting with the bot.

I felt sympathy for @lbartoletti as I found myself in a similar situation with the work having as a goal the possibility to run the testsuite w/out getting the source tree modified, as @jef-n recently mentioned on IRC, asking me: "did you give up on that work ?". Well yes, I did give up because Stale Bot pushed me to give up. If it wasn't for @jef-n asking, I would not find the power to start it again (thanks Jurgen!).

Bottom line is: Stale Bot is a _discouraging_ agent, and I reopened this because I think the contribution is valuable and it's important to keep the light on for other contributors to help reaching the goal of exposing precision.

I'd leave to @lbartoletti the task of closing it if he finds there's not point in keeping it open, but would avoid to discourage him :)

**Comment by nyalldawson** [comment]
@strk 

> But "closed" not because merged, just because a bot decided ideas have to vanish unless implemented

No -- it gets closed because it's not mergeable, and the author has abandoned the PR. An author gets plenty of notice before this happens, and it's easy enough for them to drop a comment "I'm still working on this". 

This PR definitely classifies as abandoned, and reopening is just plain noise. 

> Bottom line is: Stale Bot is a discouraging agent, and I reopened this because I think the contribution is valuable and it's important to keep the light on for other contributors to help reaching the goal of exposing precision.

Sorry to be blunt, but I completely disagree. This PR queue isn't supposed to be a dumping ground of half-finished work and wip experiments. It's for work which is READY to be reviewed and considered by the author to be in a merge-ready state. (IMO github made a mistake allowing "draft" PRs, that's a complete anti-feature to me! I would disable it on this repo if I could.).  

By allowing the PR queue to become a general dumping ground of work in all states of completeness then it becomes **impossible** for maintainers to manage. PRs which ARE good and ready to merge just sit here and decay because no-one has the time to hunt through pages of decayed requests just to find the handful of good stuff to focus on.

There's a very good reason most other mature projects have adapted a "close stale PRs aggresively" approach -- including GDAL, PROJ and yes, even GEOS. It works, and it works better than 50 pages of random bit-rotted PRs.


Back to this specific PR -- @lbartoletti will be able to clarify, but my gut feeling is that this work wasn't finished not because of stalebot, but because they just don't have the time available to make the necessary changes.

**Comment by strk** [comment]
On Wed, Jun 08, 2022 at 04:19:18PM -0700, Nyall Dawson wrote:

> Sorry to be blunt, but I completely disagree. This PR queue isn't supposed to be a dumping ground of half-finished work and wip experiments. It's for work which is READY to be reviewed and considered by the author to be in a merge-ready state. (IMO github made a mistake allowing "draft" PRs, that's a complete anti-feature to me! I would disable it on this repo if I could.).  

Let's remember at the moment opening PRs on GitHub seems to be the
ONLY WAY to verify the testsuite passes fully.

Keeping a PR open also has the ability to notify the contributor that
conflicting changes were pushed to the target branch.

> By allowing the PR queue to become a general dumping ground of work in all states of completeness then it becomes **impossible** for maintainers to manage. PRs which ARE good and ready to merge just sit here and decay because no-one has the time to hunt through pages of decayed requests just to find the handful of good stuff to focus on.

This filter seems to work:

    is:pr is:open draft:no

https://github.com/qgis/QGIS/pulls?q=is%3Apr+is%3Aopen+draft%3Ano

You can also use labels to further remove PRs waiting on feedback:

    is:pr is:open draft:no -label:Feedback

https://github.com/qgis/QGIS/issues?q=is%3Aopen+draft%3Ano+-label%3AFeedback


> There's a very good reason most other mature projects have adapted a "close stale PRs aggresively" approach -- including GDAL, PROJ and yes, even GEOS. It works, and it works better than 50 pages of random bit-rotted PRs.

Honestly I find myself fighting against all of them.
Whenever I receive the notification from stale bot (on any project) my
feeling it: "my contributions are not worth the trouble". It's really
discouraging. Time is a scarce resource; being able to share the effort
with others make the sum of scarce time become enough to finish a task
(not always, but it happens).

I think the only good reason (for a HUMAN) to close a PR is:

  1) The author agrees about it
  2) The author becomes unreachable for a long time

Last time @lbartoletti reopened the PR was 2 months ago so you may be
right that might be considered to be "unreachable for a long time" but
having been in that situation I can also undersatand if @lbartoletti
is still reachable but just sick of fighting with a bot.

> Back to this specific PR -- @lbartoletti will be able to clarify, but my gut feeling is that this work wasn't finished not because of stalebot, but because they just don't have the time available to make the necessary changes. 

I've marked this PR now with the "Feedback" label, so it can be
filtered out :)

**Review by strk — CHANGES_REQUESTED** [review]
The INSTALL.md file mention GEOS >= 3.4 is required. This change seems to instead rely on >= 3.9 without a compile-time check.

I did not verify if GEOS 3.4 would still work but I think raising a requirement should be explicitly documented, and even avoided if possible (could be just a matter of adding compile-time checks and erroring out if a grid size is specified while not supported by geos)

About the idea of a global default:  I see it valuable, but could be implemented as a second step, to change the meaning of the default value for gridSize (maybe picking a default value which does NOT have a meaning in GEOS will help with this)

**Comment by lbartoletti** [comment]
@nyalldawson @strk Some clarifications about this PR.

Orange FR has sponsored the development of these feature in GEOS years ago and use it in PostGIS. 
During a study, we showed that PostGIS contained these evolutions and that it could be interesting to have the same in QGIS, hence this PR.

I opened this PR to :
- show that we are working on it
- test the CI 
- get the first comments

But, yes, it still was a draft (missing dox, tests, etc), since we have worked on other subjects, I had not time to finish it and answer to the questions/comments. This PR is not abandonned! I prefered to let it closed, since stale bot is noisy -- but I can understand to automatically close the staled PR...

And good news, I'll work on it at the start of july -- except if @strk you want to work on it before.

**Comment by lbartoletti** [comment]
> About the idea of a global default: I see it valuable, but could be implemented as a second step, to change the meaning of the default value for gridSize (maybe picking a default value which does NOT have a meaning in GEOS will help with this)

I'm +1. 

BTW, the global default could be annoying, since one wants to have different value on some processing. I found more usable to have a value to pass on the algorithm than the same value everytime (except if you use the gui to modify it).

**Comment by strk** [comment]
On Thu, Jun 16, 2022 at 03:18:36AM -0700, Loïc Bartoletti wrote:

> And good news, I'll work on it at the start of july -- except if @strk you want to work on it before.

Great to hear! No I don't intend to work on it so you're welcome.

--strk;

File: src/core/geometry/qgsgeometryengine.h
Comment (lbartoletti) [reply src/core/geometry/qgsgeometryengine.h:115]: It's slightly different.

**Comment by lbartoletti** [comment]
> 

> @lbartoletti hi, does the grid size has an impact on snapping ?

HI @Jean-Roc ! As mentioned by phone, this PR will not impact snapping. It is used to improve the accuracy (and avoid topology errors) of overlay processing.

For information, it is already possible to snap to a grid in QGIS.

**Comment by lbartoletti** [comment]
gentle ping.

**Comment by strk** [comment]
I just thought I'd comment as when I mentioned this work to Jurgen 
he didn't know about it. Maybe this comment (by email) would reach him
if I ping @jef -- also hopefully this comment triggers posticipation
of the bot driven push-back of this contribution.

File: src/analysis/processing/qgsalgorithmdifference.cpp
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmdifference.cpp:86]: These parameters should be hidden for geos <3.9 build, and a QgsProcessingException raised if they are specified and non zero in the processAlgorithm stages on < 3.9 builds.

**Comment by jef** [comment]
@strk wrong jef 🤪

**Comment by strk** [comment]
> @strk wrong jef 🤪

Sorry, I don't understand why people need to have different
nicks on different platforms :)

Oh wait, I can !
Local identifiers are just not scalable :)

I meant @jef:osgeo.org (Matrix identifier, would Microsoft's GitHub
ever support such kind of mentions?)

**Comment by agiudiceandrea** [comment]
> I meant @jef:osgeo.org (Matrix identifier, would Microsoft's GitHub ever support such kind of mentions?)

The correct nick is @jef-n.

File: src/analysis/processing/qgsalgorithmdifference.cpp
Comment (nyalldawson) [reply src/analysis/processing/qgsalgorithmdifference.cpp:86]: Actually you can ignore this now -- we've just bumped the minimum GEOS to 3.9 so now we can be sure it's always available

Comment (lbartoletti) [reply src/analysis/processing/qgsalgorithmdifference.cpp:86]: Nice!

File: python/core/auto_generated/geometry/qgsgeometry.sip.in
Comment (nyalldawson) [inline python/core/auto_generated/geometry/qgsgeometry.sip.in:1726]:
```suggestion
:param gridSize: If this optional argument is provided, the inputs are snapped to a grid of the given size, and the result vertices are computed on that same grid (since 3.28)
```

File: python/core/auto_generated/geometry/qgsgeometry.sip.in
Comment (nyalldawson) [inline python/core/auto_generated/geometry/qgsgeometry.sip.in:1793]:
```suggestion
:param gridSize: If this optional argument is provided, the inputs are snapped to a grid of the given size, and the result vertices are computed on that same grid (since 3.28).
```

File: python/core/auto_generated/geometry/qgsgeometry.sip.in
Comment (nyalldawson) [inline python/core/auto_generated/geometry/qgsgeometry.sip.in:1821]:
```suggestion
:param gridSize: If this optional argument is provided, the inputs are snapped to a grid of the given size, and the result vertices are computed on that same grid (since 3.28).
```

Comment (nyalldawson) [reply python/core/auto_generated/geometry/qgsgeometry.sip.in:1821]: (elsewhere too)

File: src/analysis/processing/qgsalgorithmcalculateoverlaps.cpp
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmcalculateoverlaps.cpp:56]: See how the QgsLayoutAtlasToImageAlgorithm handles the "DPI" parameter -- we should do that here. It defaults to "not set" and offers the gui option to clear the value back to not set. That's cleaner than a non-self-explanatory "-1" default value.

(That'll also make it cleaner when we add a default grid size option to the processing context itself.)

File: python/core/auto_generated/geometry/qgsgeometry.sip.in
Comment (nyalldawson) [inline python/core/auto_generated/geometry/qgsgeometry.sip.in:1707]: Thought: Rather than adding a specific "gridSize" parameter to all these methods, I'd lean toward making a small "QgsGeometryOptions" class/struct which contains a gridSize member and pass that to these methods instead. We could then easily extend that class/struct with additional options in future (eg default m/z values)

**Comment by nyalldawson** [comment]
Superseded by https://github.com/qgis/QGIS/pull/50148

**Pull request closed without merging**
